prog/dftb+/lib_dftb/coordinationnumber.F90

```diff
@@ -0,0 +1,844 @@
+    real(dp), intent(in) :: k
+
+    !> Actual distance
+    real(dp), intent(in) :: r
+
+    !> Critical distance
+    real(dp), intent(in) :: r0
+
+    !> Value of the counting function in the range of [0,1]
+    real(dp) :: countFunction
+  end function countFunction
+  end interface
+
+  !> Initialize container from geometry
+  interface init
+    module procedure :: initialize
+  end interface init
+
+
+contains
+
+
+  !> Initialize coordination number container
+  subroutine initialize(self, input, nAtom, latVecs)
+
+    !> Initialised instance at return
+    type(TCNCont), intent(out) :: self
+
+    !> Nr. of atoms in the system
+    integer, intent(in) :: nAtom
+
+    !> Input for container
+    type(TCNInput), intent(in) :: input
+
+    !> Lattice vectors, if the system is periodic
+    real(dp), intent(in), optional :: latVecs(:,:)
+
+    @:ASSERT(allocated(input%covRad))
+    @:ASSERT(allocated(input%en))
+
+    self%tPeriodic = present(latVecs)
+    if (self%tPeriodic) then
+      call self%updateLatVecs(LatVecs)
+    end if
+    self%nAtom = nAtom
+
+    allocate(self%cn(nAtom))
+    allocate(self%dcndr(3, nAtom, nAtom))
+    allocate(self%dcndL(3, 3, nAtom))
+
+    self%rCutoff = input%rCutoff
+
+    select case(input%cnType)
+    case default
+      call error("Fatal programming error in dftbp_coordinationnumber!")
+    case(cnType%erf)
+      self%countFunc => erfCount
+      self%countDeriv => derfCount
+      self%kcn = 7.5_dp
+      self%tENScale = .false.
+    case(cnType%exp)
+      self%countFunc => expCount
+      self%countDeriv => dexpCount
+      self%kcn = 16.0_dp
+      self%tENScale = .false.
+    case(cnType%cov)
+      self%countFunc => erfCount
+      self%countDeriv => derfCount
+      self%kcn = 7.5_dp
+      self%tENScale = .true.
+    case(cnType%gfn)
+      self%countFunc => gfnCount
+      self%countDeriv => dgfnCount
+      self%kcn = 10.0_dp
+      self%tENScale = .false.
+    end select
+
+    self%tCutCN = input%maxCN > 0.0_dp
+    self%maxCN = input%maxCN
+
+    allocate(self%covRad(size(input%covRad)))
+    allocate(self%en(size(input%en)))
+    self%covRad(:) = input%covRad
+    self%en(:) = input%en
+
+    self%tCoordsUpdated = .false.
+
+  end subroutine initialize
+
+
+  !> Update internal stored coordinates
+  subroutine updateCoords(self, neighList, img2CentCell, coords, species0)
+
+    !> data structure
+    class(TCNCont), intent(inout) :: self
+
+    !> list of neighbours to atoms
+    type(TNeighbourList), intent(in) :: neighList
+
+    !> image to central cell atom index
+    integer, intent(in) :: img2CentCell(:)
+
+    !> atomic coordinates
+    real(dp), intent(in) :: coords(:,:)
+
+    !> central cell chemical species
+    integer, intent(in) :: species0(:)
+
+    integer, allocatable :: nNeigh(:)
+
+    allocate(nNeigh(self%nAtom))
+    call getNrOfNeighboursForAll(nNeigh, neighList, self%rCutoff)
+    call getCoordinationNumber(self%nAtom, coords, species0, nNeigh, &
```

Comment:
You can use line length up to 100 chars.

Reply:
*Can* or *must*?

Reply:
We have several old parts which are still 80 width formatted. In newer contributions I prefer 100 characters.